File: hudi-cli/src/main/java/org/apache/hudi/cli/Table.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hudi.cli;

import org.apache.hudi.common.util.Option;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Table to be rendered. This class takes care of ordering rows and limiting before renderer renders it.
 */
public class Table implements Iterable<List<String>> {

  // Header for this table
  private final TableHeader rowHeader;
  // User-specified conversions before rendering
  private final Map<String, Function<Object, String>> fieldNameToConverterMap;
  // Option attribute to track sorting field
  private final Option<String> orderingFieldNameOptional;
  // Whether sorting has to be in descending order (by default : optional)
  private final Option<Boolean> isDescendingOptional;
  // Limit the number of entries rendered
  private final Option<Integer> limitOptional;
  // Raw list of rows
  private final List<List<Comparable>> rawRows;
  // Flag to determine if all the rows have been added
  private boolean finishedAdding = false;
  // Rows ready for Rendering
  private List<List<String>> renderRows;

  public Table(TableHeader rowHeader, Map<String, Function<Object, String>> fieldNameToConverterMap,
      Option<String> orderingFieldNameOptional, Option<Boolean> isDescendingOptional, Option<Integer> limitOptional) {
    this.rowHeader = rowHeader;
    this.fieldNameToConverterMap = fieldNameToConverterMap;
    this.orderingFieldNameOptional = orderingFieldNameOptional;
    this.isDescendingOptional = isDescendingOptional;
    this.limitOptional = limitOptional;
    this.rawRows = new ArrayList<>();
  }

  /**
   * Main API to add row to the table.
   * 
   * @param row Row
   */
  public Table add(List<Comparable> row) {
    if (finishedAdding) {
      throw new IllegalStateException("Container already marked done for adding. No more entries can be added.");
    }

    if (rowHeader.getFieldNames().size() != row.size()) {
      throw new IllegalArgumentException("Incorrect number of fields in row. Expected: "
          + rowHeader.getFieldNames().size() + ", Got: " + row.size() + ", Row: " + row);
    }

    this.rawRows.add(new ArrayList<>(row));
    return this;
  }

  /**
   * Add all rows.
   * 
   * @param rows Rows to be added
   * @return
   */
  public Table addAll(List<List<Comparable>> rows) {
    rows.forEach(this::add);
    return this;
  }

  /**
   * Add all rows.
   * 
   * @param rows Rows to be added
   * @return
   */
  public Table addAllRows(List<Comparable[]> rows) {
    rows.forEach(r -> add(Arrays.asList(r)));
    return this;
  }

  /**
   * API to let the table know writing is over and reading is going to start.
   */
  public Table flip() {
    this.finishedAdding = true;
    sortAndLimit();
    return this;
  }

  /**
   * Sorting of rows by a specified field.
   * 
   * @return
   */
  private List<List<Comparable>> orderRows() {
    return orderingFieldNameOptional.map(orderingColumnName -> {
      return rawRows.stream().sorted((row1, row2) -> {
        Comparable fieldForRow1 = row1.get(rowHeader.indexOf(orderingColumnName));
        Comparable fieldForRow2 = row2.get(rowHeader.indexOf(orderingColumnName));
        int cmpRawResult = fieldForRow1.compareTo(fieldForRow2);
        return isDescendingOptional.map(isDescending -> isDescending ? -1 * cmpRawResult : cmpRawResult).orElse(cmpRawResult);
      }).collect(Collectors.toList());
    }).orElse(rawRows);
  }

  /**
   * Prepares for rendering. Rows are sorted and limited
   */
  private void sortAndLimit() {
    this.renderRows = new ArrayList<>();
    final int limit = this.limitOptional.orElse(rawRows.size());
    final List<List<Comparable>> orderedRows = orderRows();
    renderRows = orderedRows.stream().limit(limit).map(row -> IntStream.range(0, rowHeader.getNumFields()).mapToObj(idx -> {
      String fieldName = rowHeader.get(idx);
      if (fieldNameToConverterMap.containsKey(fieldName)) {
        return fieldNameToConverterMap.get(fieldName).apply(row.get(idx));
      }
      Object v = row.get(idx);
      return v == null ? "null" : v.toString();
    }).collect(Collectors.toList())).collect(Collectors.toList());
  }

  @Override
  public Iterator<List<String>> iterator() {
    if (!finishedAdding) {
      throw new IllegalStateException("Container must be flipped before reading the data");
    }
    return renderRows.iterator();
  }

  @Override
  public void forEach(Consumer<? super List<String>> action) {
    if (!finishedAdding) {
      throw new IllegalStateException("Container must be flipped before reading the data");
    }
    renderRows.forEach(action);
  }

  public List<String> getFieldNames() {
    return rowHeader.getFieldNames();
  }

  public List<List<String>> getRenderRows() {
    return renderRows;
  }
}

File: hudi-console/hudi-webapp/src/components/Table/src/types/table.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type { VNodeChild } from 'vue';
import type { PaginationProps } from './pagination';
import type { FormProps } from '/@/components/Form';
import type { TableRowSelection as ITableRowSelection } from 'ant-design-vue/lib/table/interface';
import type { ColumnProps } from 'ant-design-vue/lib/table';

import { ComponentType } from './componentType';
import { VueNode } from '/@/utils/propTypes';
import { RoleEnum } from '/@/enums/roleEnum';

export declare type SortOrder = 'ascend' | 'descend';

export interface TableCurrentDataSource<T = Recordable> {
  currentDataSource: T[];
}

export interface TableRowSelection<T = any> extends ITableRowSelection {
  /**
   * Callback executed when selected rows change
   * @type Function
   */
  onChange?: (selectedRowKeys: string[] | number[], selectedRows: T[]) => any;

  /**
   * Callback executed when select/deselect one row
   * @type Function
   */
  onSelect?: (record: T, selected: boolean, selectedRows: Object[], nativeEvent: Event) => any;

  /**
   * Callback executed when select/deselect all rows
   * @type Function
   */
  onSelectAll?: (selected: boolean, selectedRows: T[], changeRows: T[]) => any;

  /**
   * Callback executed when row selection is inverted
   * @type Function
   */
  onSelectInvert?: (selectedRows: string[] | number[]) => any;
}

export interface TableCustomRecord<T> {
  record?: T;
  index?: number;
}

export interface ExpandedRowRenderRecord<T> extends TableCustomRecord<T> {
  indent?: number;
  expanded?: boolean;
}
export interface ColumnFilterItem {
  text?: string;
  value?: string;
  children?: any;
}

export interface TableCustomRecord<T = Recordable> {
  record?: T;
  index?: number;
}

export interface SorterResult {
  column: ColumnProps;
  order: SortOrder;
  field: string;
  columnKey: string;
}

export interface FetchParams {
  searchInfo?: Recordable;
  page?: number;
  sortInfo?: Recordable;
  filterInfo?: Recordable;
  polling?: boolean;
}

export interface GetColumnsParams {
  ignoreIndex?: boolean;
  ignoreAction?: boolean;
  sort?: boolean;
}

export type SizeType = 'default' | 'middle' | 'small' | 'large';

export interface TableActionType {
  reload: (opt?: FetchParams) => Promise<void>;
  getSelectRows: <T = Recordable>() => T[];
  clearSelectedRowKeys: () => void;
  expandAll: () => void;
  expandRows: (keys: string[]) => void;
  collapseAll: () => void;
  scrollTo: (pos: string) => void; // pos: id | "top" | "bottom"
  getSelectRowKeys: () => string[];
  deleteSelectRowByKey: (key: string) => void;
  setPagination: (info: Partial<PaginationProps>) => void;
  setTableData: <T = Recordable>(values: T[]) => void;
  updateTableDataRecord: (rowKey: string | number, record: Recordable) => Recordable | void;
  deleteTableDataRecord: (rowKey: string | number | string[] | number[]) => void;
  insertTableDataRecord: (record: Recordable, index?: number) => Recordable | void;
  findTableDataRecord: (rowKey: string | number) => Recordable | void;
  getColumns: (opt?: GetColumnsParams) => BasicColumn[];
  setColumns: (columns: BasicColumn[] | string[]) => void;
  getDataSource: <T = Recordable>() => T[];
  getRawDataSource: <T = Recordable>() => T;
  setLoading: (loading: boolean) => void;
  getLoading: () => boolean | undefined;
  setProps: (props: Partial<BasicTableProps>) => void;
  redoHeight: () => void;
  setSelectedRowKeys: (rowKeys: string[] | number[]) => void;
  getPaginationRef: () => PaginationProps | boolean;
  getSize: () => SizeType;
  getRowSelection: () => TableRowSelection<Recordable>;
  getCacheColumns: () => BasicColumn[];
  emit?: EmitType;
  updateTableData: (index: number, key: string, value: any) => Recordable;
  setShowPagination: (show: boolean) => Promise<void>;
  getShowPagination: () => boolean;
  setCacheColumnsByField?: (dataIndex: string | undefined, value: BasicColumn) => void;
}

export interface FetchSetting {
  // current page field
  pageField: string;
  // page size field
  sizeField: string;
  // result field  support a.b.c
  listField: string;
  // The Total Request Results field, supported a.b.c
  totalField: string;
}

export interface TableSetting {
  redo?: boolean;
  size?: boolean;
  setting?: boolean;
  fullScreen?: boolean;
}

export interface BasicTableProps<T = any> {
  // Click on Row select
  clickToRowSelect?: boolean;
  isTreeTable?: boolean;
  // Custom sorting methods
  sortFn?: (sortInfo: SorterResult) => any;
  filterFn?: (data: Partial<Recordable<string[]>>) => any;
  // Cancel the default padding for the table
  inset?: boolean;
  showTableSetting?: boolean;
  tableSetting?: TableSetting;
  striped?: boolean;
  // Whether to automatically generate a key
  autoCreateKey?: boolean;
  // Method of calculating total rows
  summaryFunc?: (...arg: any) => Recordable[];
  // Method of calculating the total number of rows
  summaryData?: Recordable[];
  // Whether to display summary lines
  showSummary?: boolean;
  canColDrag?: boolean;
  api?: (...arg: any) => Promise<any>;
  beforeFetch?: Fn;
  afterFetch?: Fn;
  // Query criteria are processed before requesting
  handleSearchInfoFn?: Fn;
  fetchSetting?: Partial<FetchSetting>;
  // Request the interface now
  immediate?: boolean;
  // When the search form is released, whether to display the table if there is no data
  emptyDataIsShowTable?: boolean;
  // Additional request parameters
  searchInfo?: Recordable;
  defSort?: Recordable;
  useSearchForm?: boolean;
  formConfig?: Partial<FormProps>;
  columns: BasicColumn[];
  showIndexColumn?: boolean;
  indexColumnProps?: BasicColumn;
  actionColumn?: BasicColumn;
  ellipsis?: boolean;
  // Whether to inherit the parent height (parent height - form height - padding height)
  isCanResizeParent?: boolean;
  // Whether it can adapt to the height
  canResize?: boolean;
  // Adaptive height offset, calculation result - offset
  resizeHeightOffset?: number;

  // Clear options when pagination changes
  clearSelectOnPageChange?: boolean;
  //
  rowKey?: string | ((record: Recordable) => string);
  dataSource?: Recordable[];
  // Hint to the right of the title
  titleHelpMessage?: string | string[];
  // The maximum height of the table scroll
  maxHeight?: number;
  // Whether to display borders
  bordered?: boolean;
  pagination?: PaginationProps | boolean;
  loading?: boolean;

  /**
   * The column contains children to display
   * @default 'children'
   * @type string | string[]
   */
  childrenColumnName?: string;

  /**
   * Override default table elements
   * @type object
   */
  components?: object;

  /**
   * Expand all rows initially
   * @default false
   * @type boolean
   */
  defaultExpandAllRows?: boolean;

  /**
   * Initial expanded row keys
   * @type string[]
   */
  defaultExpandedRowKeys?: string[];

  /**
   * Current expanded row keys
   * @type string[]
   */
  expandedRowKeys?: string[];

  /**
   * Expanded container render for each row
   * @type Function
   */
  expandedRowRender?: (record?: ExpandedRowRenderRecord<T>) => VNodeChild | JSX.Element;

  /**
   * Customize row expand Icon.
   * @type Function | VNodeChild
   */
  expandIcon?: Function | VNodeChild | JSX.Element;

  /**
   * Whether to expand row by clicking anywhere in the whole row
   * @default false
   * @type boolean
   */
  expandRowByClick?: boolean;

  /**
   * The index of `expandIcon` which column will be inserted when `expandIconAsCell` is false. default 0
   */
  expandIconColumnIndex?: number;

  /**
   * Table footer renderer
   * @type Function | VNodeChild
   */
  footer?: Function | VNodeChild | JSX.Element;

  /**
   * Indent size in pixels of tree data
   * @default 15
   * @type number
   */
  indentSize?: number;

  /**
   * i18n text including filter, sort, empty text, etc
   * @default { filterConfirm: 'Ok', filterReset: 'Reset', emptyText: 'No Data' }
   * @type object
   */
  locale?: object;

  /**
   * Row's className
   * @type Function
   */
  rowClassName?: (record: TableCustomRecord<T>, index: number) => string;

  /**
   * Row selection config
   * @type object
   */
  rowSelection?: TableRowSelection;

  /**
   * Set horizontal or vertical scrolling, can also be used to specify the width and height of the scroll area.
   * It is recommended to set a number for x, if you want to set it to true,
   * you need to add style .ant-table td { white-space: nowrap; }.
   * @type object
   */
  scroll?: { x?: number | true; y?: number };

  /**
   * Whether to show table header
   * @default true
   * @type boolean
   */
  showHeader?: boolean;

  /**
   * Size of table
   * @default 'default'
   * @type string
   */
  size?: SizeType;

  /**
   * Table title renderer
   * @type Function | ScopedSlot
   */
  title?: VNodeChild | JSX.Element | string | ((data: Recordable) => string);

  /**
   * Set props on per header row
   * @type Function
   */
  customHeaderRow?: (column: ColumnProps, index: number) => object;

  /**
   * Set props on per row
   * @type Function
   */
  customRow?: (record: T, index: number) => object;

  /**
   * `table-layout` attribute of table element
   * `fixed` when header/columns are fixed, or using `column.ellipsis`
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/CSS/table-layout
   * @version 1.5.0
   */
  tableLayout?: 'auto' | 'fixed' | string;

  /**
   * the render container of dropdowns in table
   * @param triggerNode
   * @version 1.5.0
   */
  getPopupContainer?: (triggerNode?: HTMLElement) => HTMLElement;

  /**
   * Data can be changed again before rendering.
   * The default configuration of general user empty data.
   * You can configured globally through [ConfigProvider](https://antdv.com/components/config-provider-cn/)
   *
   * @version 1.5.4
   */
  transformCellText?: Function;

  /**
   * Callback executed before editable cell submit value, not for row-editor
   *
   * The cell will not submit data while callback return false
   */
  beforeEditSubmit?: (data: {
    record: Recordable;
    index: number;
    key: string | number;
    value: any;
  }) => Promise<any>;

  /**
   * Callback executed when pagination, filters or sorter is changed
   * @param pagination
   * @param filters
   * @param sorter
   * @param currentDataSource
   */
  onChange?: (pagination: any, filters: any, sorter: any, extra: any) => void;

  /**
   * Callback executed when the row expand icon is clicked
   *
   * @param expanded
   * @param record
   */
  onExpand?: (expande: boolean, record: T) => void;

  /**
   * Callback executed when the expanded rows change
   * @param expandedRows
   */
  onExpandedRowsChange?: (expandedRows: string[] | number[]) => void;

  onColumnsChange?: (data: ColumnChangeParam[]) => void;
}

export type CellFormat =
  | string
  | ((text: string, record: Recordable, index: number) => string | number)
  | Map<string | number, any>;

// @ts-ignore
export interface BasicColumn extends ColumnProps<Recordable> {
  children?: BasicColumn[];
  filters?: {
    text: string;
    value: string;
    children?:
      | unknown[]
      | (((props: Record<string, unknown>) => unknown[]) & (() => unknown[]) & (() => unknown[]));
  }[];

  //
  flag?: 'INDEX' | 'DEFAULT' | 'CHECKBOX' | 'RADIO' | 'ACTION';
  customTitle?: VueNode;

  slots?: Recordable;

  // Whether to hide the column by default, it can be displayed in the column configuration
  defaultHidden?: boolean;

  // Help text for table column header
  helpMessage?: string | string[];

  format?: CellFormat;

  // Editable
  edit?: boolean;
  editRow?: boolean;
  editable?: boolean;
  editComponent?: ComponentType;
  editComponentProps?:
    | ((opt: {
        text: string | number | boolean | Recordable;
        record: Recordable;
        column: BasicColumn;
        index: number;
      }) => Recordable)
    | Recordable;
  editRule?: boolean | ((text: string, record: Recordable) => Promise<string>);
  editValueMap?: (value: any) => string;
  onEditRow?: () => void;
  // The permission encoding controls whether it is displayed
  auth?: RoleEnum | RoleEnum[] | string | string[];
  // Whether business controls are displayed
  ifShow?: boolean | ((column: BasicColumn) => boolean);
  // Customize what is displayed after modifications
  editRender?: (opt: {
    text: string | number | boolean | Recordable;
    record: Recordable;
    column: BasicColumn;
    index: number;
  }) => VNodeChild | JSX.Element;
}

export type ColumnChangeParam = {
  dataIndex: string;
  fixed: boolean | 'left' | 'right' | undefined;
  visible: boolean;
};

export interface InnerHandlers {
  onColumnsChange: (data: ColumnChangeParam[]) => void;
}
